src/components/ProductCard.tsx

```diff
@@ -3,13 +3,26 @@ import {
   Card,
   Button,
 } from 'react-bootstrap';
+import { connect, ConnectedProps } from 'react-redux'
+import { addToCart } from '../actions/cart';
 import { Product } from '../types/product';
 
-type Props = {
+const mapDispatch = {
```

Comment:
Since this is a functional component, it's more common to use the `useDispatch` hook to dispatch a request. This has the benefit of not needing to wrap a component in a HoC, and tends to be easier to get correct types for (one the initial setup is done).

For example:
```ts
const dispatch = useDispatch()
dispatch(ActionCreator());
```

`useDispatch` also takes in a generic which allows the dispatched actions to be type checked. Internally at FS, we use these type helpers to accomplish this (slightly different). I think it's really hard to have good types with Redux without these helpers, and a similar open source library (that we drew inspiration from) is [typesafe-actions](https://github.com/piotrwitek/typesafe-actions).

```ts
import { Action } from "redux";

export type ActionCreator<T extends string> = (...args: any[]) => Action<T>;
export type ActionType<AC> = 
  AC extends ActionCreator<string> ?
  ReturnType<AC> :
  AC extends Record<string, any> ?
    { [K in keyof AC]: ActionType<AC[K]> }[keyof AC]
    : never;
```

The `ActionType` is a recursive generic that supports finding the types of an individual ActionCreator, or a dictionary (can be nested). 

This has to be paired with the `createAction` helper on the actions. The `T extends string` portion of the generic allows typescript to understand the type strings as constants.

```
export const createAction = <T extends string, A extends Action<T>>(action: A): A => action;
```

We need to extend how the types are generated for the actions:
```ts
import * as CartActions from './cart';
export const Actions = {
  ...CartActions,
}

export type AppActions = ActionType<typeof Actions>;
export type AppDispatch = Dispatch<AppActions>;
```
When applied to the actions, you can see that AppActions now gives strong union types that automatically update whenever the action creators are updated:
![image](https://user-images.githubusercontent.com/1714798/76648215-a4742000-6534-11ea-9a1b-48e144792072.png)

We can use the AppDispatch to type useDispatch, to ensure that we're only dispatching the correct events:
```ts
  const dispatch = useDispatch<AppDispatch>();
```

This.. is a lot. I've got some local changes that I'll push up to a branch so you can see it in action.

Reply:
https://github.com/fullstorydev/react-shoppe-demo/pull/8

Reply:
> #8

Amazing! Thanks Tim. A little foreshadowing for PRs to come: I'm currently relying on the HOC to map global state to the component as well. I'd love to know how "map state to props" works without an HOC.

Reply:
Per our slack chat - I'll move forward with `useDispatch` and I'll revisit the action helpers in the future.

Reply:
OK! I've refactored to use `useDispatch`. Per our slack chat: I've retained `createAction` but have left out the other Action helper types (for now). I've removed all explicitly defined cart action types.